connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java

```diff
@@ -0,0 +1,189 @@
+package com.google.cloud.spark.bigquery;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.FieldList;
+import com.google.cloud.bigquery.LegacySQLTypeName;
+import org.apache.avro.generic.GenericRecord;
+import com.google.cloud.bigquery.Schema;
+import org.apache.avro.util.Utf8;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.util.GenericArrayData;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+public class SchemaConverters {
+    // Numeric is a fixed precision Decimal Type with 38 digits of precision and 9 digits of scale.
+    // See https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#numeric-type
+    private final static int BQ_NUMERIC_PRECISION = 38;
+    private final static int BQ_NUMERIC_SCALE = 9;
+    private final static DecimalType NUMERIC_SPARK_TYPE = DataTypes.createDecimalType(
+    BQ_NUMERIC_PRECISION, BQ_NUMERIC_SCALE);
+
+    /** Convert a BigQuery schema to a Spark schema */
+    public static StructType toSpark(Schema schema) {
+        List<StructField> fieldList = schema.getFields().stream()
+                .map(SchemaConverters::convert).collect(Collectors.toList());
+        StructType structType = new StructType(fieldList.toArray(new StructField[0]));
+
+        return structType;
+    }
+
+    public static InternalRow createRowConverter(Schema schema, List<String> namesInOrder, GenericRecord record) {
+        return convertAll(schema.getFields(), record, namesInOrder);
+    }
+
+    static Object convert(Field field, Object value) {
+        if (value == null) {
+            return null;
+        }
+
+        if (field.getMode() == Field.Mode.REPEATED) {
+            // rather than recurring down we strip off the repeated mode
+            // Due to serialization issues, reconstruct the type using reflection:
+            // See: https://github.com/googleapis/google-cloud-java/issues/3942
```

Comment:
@davidrabinowitz Scala used "Some" and "Option" annotations here. I am not sure how that would be applicable in Java

Reply:
Originally they are used in the `convert(Field field)` method, not in this one - and they are not needed in Java. Do you have another place where those are used?

Reply:
Not in this file.